src/mjolnir/graphenhancer.cc

```diff
@@ -187,40 +187,327 @@ void UpdateSpeed(DirectedEdge& directededge, const uint32_t density, const uint3
   }
 }
 
+// Get the turn types.  This is used the determine if we should enhance or update
+// Turn lanes based on the turns at this node.
+void GetTurnTypes(const DirectedEdge directededge,
+                  const uint32_t idx,
+                  std::set<Turn::Type>& outgoing_turn_type,
+                  NodeInfo& startnodeinfo,
+                  GraphTileBuilder& tilebuilder,
+                  GraphReader& reader,
+                  std::mutex& lock) {
+
+  // Get the tile at the startnode
+  const GraphTile* tile = &tilebuilder;
+  uint32_t heading = startnodeinfo.heading(idx);
+
+  // Get the tile at the end node. and find inbound heading of the candidate
+  // edge to the end node.
+  if (tile->id() != directededge.endnode().Tile_Base()) {
+    lock.lock();
+    tile = reader.GetGraphTile(directededge.endnode());
+    lock.unlock();
+  }
+  const NodeInfo* node = tile->node(directededge.endnode());
+
+  const DirectedEdge* diredge = tile->directededge(node->edge_index());
+  for (uint32_t i = 0; i < node->edge_count(); i++, diredge++) {
+    // Find the opposing directed edge and its heading
+    if (i == directededge.opp_local_idx()) {
+      auto shape = tile->edgeinfo(diredge->edgeinfo_offset()).shape();
+      if (!diredge->forward()) {
+        std::reverse(shape.begin(), shape.end());
+      }
+      uint32_t hdg = std::round(
+          PointLL::HeadingAlongPolyline(shape, GetOffsetForHeading(diredge->classification(),
+                                                                   diredge->use())));
+
+      // Convert to inbound heading
+      heading = ((hdg + 180) % 360);
+      break;
+    }
+  }
+
+  // Iterate through outbound edges and get turn degrees from the candidate
+  // edge onto outbound driveable edges.
+  diredge = tile->directededge(node->edge_index());
+  for (uint32_t i = 0; i < node->edge_count(); i++, diredge++) {
+    // Skip opposing directed edge and any edge that is not a road. Skip any
+    // edges that are not driveable outbound.
+    if (i == directededge.opp_local_idx() || !(diredge->forwardaccess() & kAutoAccess) ||
+        (directededge.restrictions() & (1 << diredge->localedgeidx())) != 0) {
+      continue;
+    }
+
+    // Get the heading of the outbound edge (unfortunately GraphEnhancer may
+    // not have yet computed and stored headings for this node).
+    auto shape = tile->edgeinfo(diredge->edgeinfo_offset()).shape();
+    if (!diredge->forward()) {
+      std::reverse(shape.begin(), shape.end());
+    }
+    uint32_t to_heading =
+        std::round(PointLL::HeadingAlongPolyline(shape, GetOffsetForHeading(diredge->classification(),
+                                                                            diredge->use())));
+    // Store outgoing turn type for any driveable edges
+    uint32_t turndegree = GetTurnDegree(heading, to_heading);
+    outgoing_turn_type.insert(Turn::GetType(turndegree));
+  }
+}
+
+// Update the rightmost lane as needed.
```

Comment:
This is where we enhance the lanes by adding right lane information where the user added none.  We only add slight rights when the through lanes have be accounted for in the turn types.